src/librvnpal/src/win/fileutils.c

```diff
@@ -0,0 +1,158 @@
+#include <windows.h>
+#include <stdint.h>
+#include <assert.h>
+
+#include "rvn.h"
+#include "status_codes.h"
+#include "internal_win.h"
+
+PRIVATE int32_t
+_resize_file(void *handle, int64_t size, int32_t *detailed_error_code)
+{
+    int32_t rc;
+    LARGE_INTEGER distance_to_move;
+    distance_to_move.QuadPart = size;
+    if (SetFilePointerEx(handle, distance_to_move, NULL, FILE_BEGIN) == FALSE)
+    {
+        rc = FAIL_SET_FILE_POINTER;
+        goto error_cleanup;
+    }
+
+    if (SetEndOfFile(handle) == FALSE)
+    {
+        rc = FAIL_SET_EOF;
+        goto error_cleanup;
+    }
+
+    return SUCCESS;
+
+error_cleanup:
+    *detailed_error_code = GetLastError();
+    return rc;
+}
+
+PRIVATE int32_t
+_write_file_incrementally(void* handle, char** buffer, int64_t* size, int64_t* offset, int64_t increment_size, int32_t* detailed_error_code)
+{
+    OVERLAPPED overlapped;
+    memset(&overlapped, 0, sizeof(overlapped));
+    overlapped.Offset = (int)(*offset & 0xffffffff);
+    overlapped.OffsetHigh = (int)(*offset >> 32);
+
+    DWORD actual_size_to_write;
+    while (*size > 0)
+    {
+        if(*size < increment_size)
+        {
+            actual_size_to_write = *size;
+        }
+        else
+        {
+            actual_size_to_write = increment_size;
+        }
+
+        if (WriteFile(handle, buffer, actual_size_to_write, NULL, &overlapped) == FALSE)
+        {
+            *detailed_error_code = GetLastError();
+            return FAIL_WRITE_FILE;
+        }
+
+        *buffer += actual_size_to_write;
+        *size -= actual_size_to_write;
+
+        *offset += actual_size_to_write;
+        overlapped.Offset = (int)(*offset & 0xffffffff);
+        overlapped.OffsetHigh = (int)(*offset >> 32);
+    }
+
+    return SUCCESS;
+}
+
+PRIVATE int32_t
+_write_file(void* handle, const void* buffer, int64_t size, int64_t offset, int32_t* detailed_error_code)
+{
+    const int32_t WRITE_INCREMENT = 4096;
+
+    assert(size % WRITE_INCREMENT == 0);
+    assert((int64_t)buffer % WRITE_INCREMENT == 0);
+
+    int32_t rc = _write_file_incrementally(handle, (char**)&buffer, &size, &offset, UINT32_MAX, detailed_error_code);
+    if (rc == SUCCESS)
+        return SUCCESS;
+
+    if (*detailed_error_code != ERROR_WORKING_SET_QUOTA)
+        return FAIL_WRITE_FILE;
+
+    // this error can happen under low memory conditions, instead of trying to write the whole thing in a single shot
+    // we'll write it in 4KB increments. This is likely to be much slower, but failing here will fail the entire DB
+    return _write_file_incrementally(handle, (char**)&buffer, &size, &offset, WRITE_INCREMENT, detailed_error_code);
```

Comment:
I _think_ that you are trying to continue from the previous location, but that isn't what we want.
If we failed, we assume that we failed completely and do this from the beginning

Reply:
But we know where we stopped why not to start from there?

Reply:
So? You are already in error state, better to make sure.
And the real issue is that I want to make sure that the code is simpler. Better than having to deal with very rare event and complicate matters.